<a href="https://colab.research.google.com/github/FishchevaKM/MachineLearningHomework/blob/main/FirstLab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Каждый вид идентифицируется как съедобный, ядовитый или неизвестной съедобности и не рекомендуется, кроме того присутствуют описательные характеристики:


1.	class - класс
2.	cap shape - форма шляпки
3.	cap surface - поверхность шляпки
4.	cap color - цвет шляпки
5.	bruises - повреждения
6.	odor - запах
7.	gill attachment - жаберное крепление
8.	gill spacing – плотность жабр
9.	gill size – размер жабр
10.	gill color – цвет жабр
11.	stalk shape - форма стебля
12.	stalk root - корень стебля
13.	stalk surface above ring - поверхность стебля над кольцом
14.	stalk surface below ring - поверхность стебля под кольцом
15.	stalk color above ring - цвет стебля над кольцом
16.	stalk color below ring - цвет стебля под кольцом
17.	veil type – тип покрова
18.	veil color – цвет  покрова
19.	ring number – число колец
20.	ring type – тип колец
21.	spore print color – цвет спор
22.	population - популяция
23.	habitat -  ареал


In [3]:
import numpy as np              # Массивы (матрицы, векторы, линейная алгебра)
import matplotlib.pyplot as plt # Научная графика
%matplotlib inline 
    # Говорим jupyter'у, чтобы весь графический вывод был в браузере, а не в отдельном окне
import pandas as pd             # Таблицы и временные ряды (dataframe, series)
import seaborn as sns           # Еще больше красивой графики для визуализации данных
import sklearn                  # Алгоритмы машинного обучения

Считываем данные

In [4]:
url = "https://raw.githubusercontent.com/FishchevaKM/MachineLearningHomework/main/MushroomBase.csv"
data = pd.read_csv(url, sep = ";")

Таблица включает описания гипотетических образцов, соответствующих 23 видам гриба семейства Agaricus и Lepiota, взятые из полевого справочника по североамериканским грибам Одюбонского Общества (1981). Каждый вид идентифицирован как определенно съедобный, определенно ядовитый или обладающий неизвестной съедобностью и не рекомендуемый. Этот последний класс был объединен с ядовитым. В Руководстве четко указано, что не существует простого правила определения съедобности гриба;

Вначале посмотрим на размеры таблицы - количество строк (каждая строка соответствует одному грибу) и количество столбцов (каждый столбец соответствует конкретному признаку):

In [5]:
data

Class Cap-shape Cap-surface  ... Spore-print-color Population Habitat
0        p         x           s  ...                 k          s       u
1        e         x           s  ...                 n          n       g
2        e         b           s  ...                 n          n       m
3        p         x           y  ...                 k          s       u
4        e         x           s  ...                 n          a       g
...    ...       ...         ...  ...               ...        ...     ...
8119     e         k           s  ...                 b          c       l
8120     e         x           s  ...                 b          v       l
8121     e         f           s  ...                 b          c       l
8122     p         k           y  ...                 w          v       l
8123     e         x           s  ...                 o          c       l

[8124 rows x 23 columns]

Вывод первых 5-ти строк

In [6]:
data.head()

Class Cap-shape Cap-surface  ... Spore-print-color Population Habitat
0     p         x           s  ...                 k          s       u
1     e         x           s  ...                 n          n       g
2     e         b           s  ...                 n          n       m
3     p         x           y  ...                 k          s       u
4     e         x           s  ...                 n          a       g

[5 rows x 23 columns]

В полученой таблице, например, гриб №3 имеет характеристики:

1.	Class – класс: p(poisonous)
2.	cap shape – форма шляпки: x(convex-)
3.	cap surface - поверхность шляпки: y(scaly)
4.	cap color - цвет шляпки: w(white)
5.	bruises – повреждения: t(true)
6.	odor – запах: p(pungent)
7.	gill attachment - жаберное крепление: f(free)
8.	gill spacing – плотность жабр: c(close)
9.	gill size – размер жабр: n(narrow)
10.	gill color – цвет жабр: n(red)
11.	stalk shape - форма стебля: e(enlarging)
12.	stalk root - корень стебля: e(equal)
13.	stalk surface above ring - поверхность стебля над кольцом: s(smooth)
14.	stalk surface below ring - поверхность стебля под кольцом: s(smooth)
15.	stalk color above ring - цвет стебля над кольцом: w(white)
16.	stalk color below ring - цвет стебля под кольцом: w(white)
17.	veil type – тип покрова: p(partial)
18.	veil color – цвет  покрова: w(white)
19.	ring number – число колец: o(one)
20.	ring type – тип колец: p(pendant)
21.	spore print color – цвет спор: k(black)
22.	population – популяция: s(scattered)
23.	habitat -  ареал: u(urban)


Вывод последних 5-и строк

In [7]:
data.tail()

Class Cap-shape Cap-surface  ... Spore-print-color Population Habitat
8119     e         k           s  ...                 b          c       l
8120     e         x           s  ...                 b          v       l
8121     e         f           s  ...                 b          c       l
8122     p         k           y  ...                 w          v       l
8123     e         x           s  ...                 o          c       l

[5 rows x 23 columns]

In [11]:
data.shape

(8124, 23)

В данной таблице 23 столбца и 8124 строки (23 характеристики, 8124 гриба).

Можно посмотреть имеющиеся названия столбцов с помощью

In [12]:
data.columns

Index(['Class', 'Cap-shape', 'Cap-surface', 'Cap-color', 'Bruises', 'Odor',
       'Gill-attachment', 'Gill-spacing', 'Gill-size', 'Gill-color',
       'Stalk-shape', 'Stalk-root', 'Stalk-surface-above-ring',
       'Stalk-surface-below-ring', 'Stalk-color-above-ring',
       'Stalk-color-below-ring', 'Veil-type', 'Veil-color', 'Ring-number',
       'Ring-type', 'Spore-print-color', 'Population', 'Habitat'],
      dtype='object')

In [13]:
data.index

RangeIndex(start=0, stop=8124, step=1)

Или же посмотреть значение выбранного столбца в конкретной строке таблице.

In [15]:
data.at[8122, 'Odor']

'y'

Вывод выбранного столбца.

In [18]:
data['Odor']

0       p
1       a
2       l
3       p
4       n
       ..
8119    n
8120    n
8121    n
8122    y
8123    n
Name: Odor, Length: 8124, dtype: object

In [19]:
data['Odor'][3]

'p'

Узнаем общую информацию по нашей таблице.

In [22]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Class                     8124 non-null   object
 1   Cap-shape                 8124 non-null   object
 2   Cap-surface               8124 non-null   object
 3   Cap-color                 8124 non-null   object
 4   Bruises                   8124 non-null   object
 5   Odor                      8124 non-null   object
 6   Gill-attachment           8124 non-null   object
 7   Gill-spacing              8124 non-null   object
 8   Gill-size                 8124 non-null   object
 9   Gill-color                8124 non-null   object
 10  Stalk-shape               8124 non-null   object
 11  Stalk-root                5644 non-null   object
 12  Stalk-surface-above-ring  8124 non-null   object
 13  Stalk-surface-below-ring  8124 non-null   object
 14  Stalk-color-above-ring  

Вывод первых пяти элементов указанных нами столбцов.

In [23]:
data[['Class', 'Population']].head()

Class Population
0     p          s
1     e          n
2     e          n
3     p          s
4     e          a

Выведем основные статистические характеристики данных нашей таблицы.

In [46]:
data.describe()

Class Cap-shape Cap-surface  ... Ring-number Ring-type Spore-print-color
count   8124      8124        8124  ...        8124      8124              8124
unique     2         6           4  ...           3         5                 9
top        e         x           y  ...           o         p                 w
freq    4208      3656        3244  ...        7488      3968              2388

[4 rows x 20 columns]

где count – количество элементов в каждом столбце; unique – количество уникальных значений в столбце; top – наиболее часто встречающееся значение; freq – количество top значения в столбце.

In [41]:
data.corr

<bound method DataFrame.corr of      Class Cap-shape Cap-surface  ... Ring-number Ring-type Spore-print-color
0        p         x           s  ...           o         p                 k
1        e         x           s  ...           o         p                 n
2        e         b           s  ...           o         p                 n
3        p         x           y  ...           o         p                 k
4        e         x           s  ...           o         e                 n
...    ...       ...         ...  ...         ...       ...               ...
8119     e         k           s  ...           o         p                 b
8120     e         x           s  ...           o         p                 b
8121     e         f           s  ...           o         p                 b
8122     p         k           y  ...           o         e                 w
8123     e         x           s  ...           o         p                 o

[8124 rows x 20 columns]>

Проверим элементы на соответствие заданной характеристике.

In [25]:
data['Class'] == 'e'

0       False
1        True
2        True
3       False
4        True
        ...  
8119     True
8120     True
8121     True
8122    False
8123     True
Name: Class, Length: 8124, dtype: bool

Грибы со значением true являются съедобными, а со значением false – несъедобными.

Удалим столбцы, которые никак не помогут нам определить съедобность гриба.

In [26]:
data = data.drop('Bruises', 1)

In [30]:
data = data.drop('Population', 1)

KeyError: ignored

In [29]:
data = data.drop('Habitat', 1)

KeyError: ignored

Готовим данные для дальнейшей работы

Узнаем категорию данных одного столбца.

In [34]:
data['Ring-type'].dtype.name

'object'

Разделим имеющиеся столбцы на категориальные и на числовые.

In [32]:
categorical_columns = [c for c in data.columns if data[c].dtype.name == 'object']
numerical_columns   = [c for c in data.columns if data[c].dtype.name != 'object']
print(categorical_columns)
print(numerical_columns)

['Class', 'Cap-shape', 'Cap-surface', 'Cap-color', 'Odor', 'Gill-attachment', 'Gill-spacing', 'Gill-size', 'Gill-color', 'Stalk-shape', 'Stalk-root', 'Stalk-surface-above-ring', 'Stalk-surface-below-ring', 'Stalk-color-above-ring', 'Stalk-color-below-ring', 'Veil-type', 'Veil-color', 'Ring-number', 'Ring-type', 'Spore-print-color']
[]


В моей таблице все столбцы являются категориальными.

Для каждого столбца определим возможные значения.

In [33]:
for c in categorical_columns:
    print(c, data[c].unique())

Class ['p' 'e']
Cap-shape ['x' 'b' 's' 'f' 'k' 'c']
Cap-surface ['s' 'y' 'f' 'g']
Cap-color ['n' 'y' 'w' 'g' 'e' 'p' 'b' 'u' 'c' 'r']
Odor ['p' 'a' 'l' 'n' 'f' 'c' 'y' 's' 'm']
Gill-attachment ['f' 'a']
Gill-spacing ['c' 'w']
Gill-size ['n' 'b']
Gill-color ['k' 'n' 'g' 'p' 'w' 'h' 'u' 'e' 'b' 'r' 'y' 'o']
Stalk-shape ['e' 't']
Stalk-root ['e' 'c' 'b' 'r' nan]
Stalk-surface-above-ring ['s' 'f' 'k' 'y']
Stalk-surface-below-ring ['s' 'f' 'y' 'k']
Stalk-color-above-ring ['w' 'g' 'p' 'n' 'b' 'e' 'o' 'c' 'y']
Stalk-color-below-ring ['w' 'p' 'g' 'b' 'n' 'e' 'y' 'o' 'c']
Veil-type ['p']
Veil-color ['w' 'n' 'o' 'y']
Ring-number ['o' 't' 'n']
Ring-type ['p' 'e' 'l' 'f' 'n']
Spore-print-color ['k' 'n' 'u' 'h' 'w' 'r' 'o' 'y' 'b']


Обрабатываем пропущенные значения

In [35]:
data[categorical_columns].count(axis = 0)

Class                       8124
Cap-shape                   8124
Cap-surface                 8124
Cap-color                   8124
Odor                        8124
Gill-attachment             8124
Gill-spacing                8124
Gill-size                   8124
Gill-color                  8124
Stalk-shape                 8124
Stalk-root                  5644
Stalk-surface-above-ring    8124
Stalk-surface-below-ring    8124
Stalk-color-above-ring      8124
Stalk-color-below-ring      8124
Veil-type                   8124
Veil-color                  8124
Ring-number                 8124
Ring-type                   8124
Spore-print-color           8124
dtype: int64

В столбце Stalk-root есть пропущенные значения. Заменим их на самое популярное значение в этом столбце.

In [36]:
data['Stalk-root'].describe().top

'b'

In [43]:
data['Stalk-root'] = data['Stalk-root'].fillna('b')

Еще раз проверим пропущенные значения:

In [44]:
data[categorical_columns].count(axis = 0)

Class                       8124
Cap-shape                   8124
Cap-surface                 8124
Cap-color                   8124
Odor                        8124
Gill-attachment             8124
Gill-spacing                8124
Gill-size                   8124
Gill-color                  8124
Stalk-shape                 8124
Stalk-root                  8124
Stalk-surface-above-ring    8124
Stalk-surface-below-ring    8124
Stalk-color-above-ring      8124
Stalk-color-below-ring      8124
Veil-type                   8124
Veil-color                  8124
Ring-number                 8124
Ring-type                   8124
Spore-print-color           8124
dtype: int64

In [47]:
data.describe

<bound method NDFrame.describe of      Class Cap-shape Cap-surface  ... Ring-number Ring-type Spore-print-color
0        p         x           s  ...           o         p                 k
1        e         x           s  ...           o         p                 n
2        e         b           s  ...           o         p                 n
3        p         x           y  ...           o         p                 k
4        e         x           s  ...           o         e                 n
...    ...       ...         ...  ...         ...       ...               ...
8119     e         k           s  ...           o         p                 b
8120     e         x           s  ...           o         p                 b
8121     e         f           s  ...           o         p                 b
8122     p         k           y  ...           o         e                 w
8123     e         x           s  ...           o         p                 o

[8124 rows x 20 columns]>

In [48]:
data_describe = data.describe(include = [object])
for c in categorical_columns:
    data[c] = data[c].fillna(data_describe[c]['top'])